### PR TITLE
Aria2 jsonrpc support

### LINKED ISSUE #3460: aria2.py  is not compatible to new version of aira2
### Steps to reproduce:
- Step 1: ...

#### Config:
```yaml
Paste FULL config and remove any personal info if config is too long, attach the file to the ticket.
If issue is with a single task, you can get get resulting configuration by running:
  flexget execute --task <NAME> --dump-config
Make sure to redact any personal information (passwords, api keys, etc) !
```
  
#### Log:
<details>
 <summary>(click to expand)</summary>

```
paste log output here
```
</details>

### Additional information:

- FlexGet version:
- Python version:
- Installation method:
- Using daemon (yes/no):
- OS and version:
- Link to crash log:

hi, good gays, thanks for your good tools.
now i have a problem, the new  aria2  RPC  could not support http,  only support https/wss,  i connect  https://xxx:jsonrpc well in aria2ng.  but it give the error message below in flexget.  thanks a lot!
INFO     aria2         soft            Connecting to daemon at **### http://xxxx:2083/rpc**
CRITICAL task          soft            BUG: Unhandled error in plugin aria2: <ProtocolError for **### xxx:2083/rpc: 400 Bad Request**>

### PR DESCRIPTION
**Pull request by RyoLee**
### Motivation for changes:
- Add support  for JSON-RPC 

### Detailed changes:
- Add configs: 
  - scheme: http(default)/https
  - rpc_mode: xml(default)/json
  - rpc_path: rpc(default)
- JSON-RPC client

### Addressed issues:
- Fixes #3460 .

### Implemented feature requests:
- N/A

### Config usage if relevant (new plugin or updated schema):
```
web_server:
  bind: 0.0.0.0
  port: 5050
  web_ui: yes

tasks:
  SPYxFAMILY:
    rss: https://******
    accept_all: yes
    aria2:
        server: ******
        port: ******
        scheme: https
        secret: ******
        rpc_mode: json
        rpc_path: jsonrpc
        path: /downloads/Flexget
schedules:
  - tasks: '*'
    interval:
      minutes: 10
```
### Log and/or tests output (preferably both):
flexget:
```
2022-07-31 03:46:43 INFO     web_server_daemon                 Running web server at IP 0.0.0.0:5050
2022-07-31 03:46:43 INFO     web_server_daemon                 Initiating API
2022-07-31 03:46:43 INFO     web_server_daemon                 Registering WebUI v2
2022-07-31 03:46:43 INFO     web_server                    Web server started at http://10.42.0.102:5050
2022-07-31 03:46:43 INFO     web_server                    API available at http://10.42.0.102:5050/api
2022-07-31 03:46:43 INFO     scheduler                     Starting scheduler
2022-07-31 03:46:43 INFO     web_server                    WebUI (v2) available at http://10.42.0.102:5050/
2022-07-31 03:46:49 INFO     aria2         SPYxFAMILY      Connecting to daemon at https://******:******/jsonrpc
2022-07-31 03:49:05 INFO     aria2         SPYxFAMILY      Connecting to daemon at https://******:******/jsonrpc
2022-07-31 03:59:10 INFO     aria2         SPYxFAMILY      Connecting to daemon at https://******:******/jsonrpc
```
aria2:
```
07/31 03:46:51 [NOTICE] Serialized session to '/config/aria2.session' successfully.

07/31 03:46:51 [NOTICE] Download complete: /downloads/Flexget/[诸神字幕组][间谍过家家][Spy x Family][12][简繁日语字幕][1080P][MKV HEVC].torrent

07/31 03:46:51 [NOTICE] Download complete: /downloads/Flexget/[诸神字幕组][间谍过家家][Spy x Family][10][简繁日语字幕][1080P][MKV HEVC].torrent
07/31 03:46:51 [WARRING] Skip delete. Task status invalid: complete
07/31 03:46:51 [WARRING] Skip delete. Task status invalid: complete
07/31 03:46:51 [INFO] General download task, skipped delete .torrent file.
07/31 03:46:51 [INFO] Deleting empty directory ...
07/31 03:46:51 [INFO] General download task, skipped delete .torrent file.
07/31 03:46:51 [INFO] Deleting empty directory ...

07/31 03:46:51 [NOTICE] Download complete: /downloads/Flexget/[诸神字幕组][间谍过家家][Spy x Family][09][简繁日语字幕][1080P][MKV HEVC].torrent
07/31 03:46:51 [WARRING] Skip delete. Task status invalid: complete
07/31 03:46:51 [INFO] General download task, skipped delete .torrent file.
07/31 03:46:51 [INFO] Deleting empty directory ...
```
#### To Do:

- [x] ~~Testing requirements: XML-RPC client (I am using a custom version of Aria2, which prevents me from testing the XML-RPC client)~~ Tested
- [ ] WS(and WSS) protocol support

### CODE REVIEW

File: flexget/plugins/clients/aria2.py [hunk 1]
```diff
@@ -12,6 +15,135 @@
...
+    ADDURI_METHOD = 'aria2.addUri'
+    GETGLOBALSTAT_METHOD = 'aria2.getGlobalStat'
+    ADDTORRENT_METHOD = 'aria2.addTorrent'
+    ADDMETALINK_METHOD = 'aria2.addMetalink'
+
+    def __init__(self, server, port, scheme, rpc_path, username=None, password=None, token=None):
+        if token:
+            self._token = 'token:' + token
+        else:
+            self._token = token
```
Comment (ianstalk) [inline flexget/plugins/clients/aria2.py:32]: Why isn't this set to empty string like userpass?

Comment (RyoLee) [reply flexget/plugins/clients/aria2.py:32]: both ```None``` and  ```''``` are correct, it's equivalent at [line50](https://github.com/Flexget/Flexget/pull/3515/files/587cacd355c4a4ce5d42bebd14f93200c8e56c2b#diff-86d45d970d5f22e4eb4842061c85f66e512d96bc34a8de8649e1ea4f10f4fe9dR50)
do I need to change it to an empty string?

Comment (ianstalk) [reply flexget/plugins/clients/aria2.py:32]: I think it would be clearer and more consistent.

Comment (RyoLee) [reply flexget/plugins/clients/aria2.py]: updated https://github.com/Flexget/Flexget/pull/3515/commits/aa86ea9f0c229fafac8c4756dec123b17a382742

File: flexget/plugins/clients/aria2.py [hunk 2]
```diff
@@ -12,6 +15,135 @@
...
+class XMLRPC(object):
+    def __init__(self, server, port, scheme, rpc_path, username=None, password=None, token=None):
+        if token:
+            self._token = 'token:' + token
+        else:
+            self._token = ''
+        if username and password:
+            userpass = '%s:%s@' % (username, password)
+        else:
+            userpass = ''
```
Comment (ianstalk) [inline flexget/plugins/clients/aria2.py:97]: This code is duplicated across both methods. maybe make a helper function to do it?

File: flexget/plugins/clients/aria2.py [hunk 3]
```diff
@@ -12,6 +15,135 @@
 logger = logger.bind(name='aria2')
 
 
+class JSONRPC(object):
```
[the remark this answers is not in the document]
Comment (RyoLee) [reply flexget/plugins/clients/aria2.py]: refactored https://github.com/Flexget/Flexget/pull/3515/commits/187f8b28313a66b34862ec7965c08e4f92dfc817
auto fix (code style) https://github.com/Flexget/Flexget/pull/3515/commits/17ed29f3430a8210f6f9e67bf224bf2478512d09

File: flexget/plugins/clients/aria2.py [hunk 4]
```diff
@@ -12,6 +15,138 @@
...
+        logger.critical('Fault code {} message {}', code, message)
+        raise plugin.PluginError('Fault code %s message %s' % (code, message), logger)
+
+    def _default_success_handle(response):
+        return response.text
+
+    def _post(
+        self, method, params, on_success=_default_success_handle, on_fail=_default_error_handle
+    ):
+        _params = self._get_req_params(method, params)
```
Comment (paranoidi) [inline flexget/plugins/clients/aria2.py:77]: variables should not usually be prefixed with _, just inline the variable?

File: flexget/plugins/clients/aria2.py [hunk 5]
```diff
@@ -12,6 +15,138 @@
...
+        return self._post(JsonRpcClient.ADDTORRENT_METHOD, params=[torrent, options])
+
+    def get_global_stat(self):
+        # https://aria2.github.io/manual/en/html/aria2c.html#aria2.getGlobalStat
+        return self._post(JsonRpcClient.GETGLOBALSTAT_METHOD, params=[])
+
+
+class XmlRpcClient(RpcClient):
+    def __init__(self, server, port, scheme, rpc_path, username=None, password=None, token=None):
+        _schemes = {'http': None, 'https': ssl.SSLContext()}
```
Comment (paranoidi) [inline flexget/plugins/clients/aria2.py:101]: _prefixed variable

**Review by paranoidi — APPROVED** [review]
Looks good, just couple style points

**Comment by paranoidi** [comment]
Thanks for taking time to do this! Much appreciated

**Pull request merged**